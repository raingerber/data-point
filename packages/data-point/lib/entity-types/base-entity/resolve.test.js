/* eslint-env jest */
'use strict'

const ResolveEntity = require('./resolve')
const ReducerEntity = require('../../reducer-types/reducer-entity')
const ReducerExpression = require('../../reducer-types/reducer-expression')

const FixtureStore = require('../../../test/utils/fixture-store')
const helpers = require('../../helpers')
const utils = require('../../utils')

let dataPoint
let resolveTransform

beforeAll(() => {
  dataPoint = FixtureStore.create()
  resolveTransform = helpers.createResolveTransform(dataPoint)
})

afterEach(() => {
  dataPoint.middleware.clear()
})

describe('ResolveEntity.resolveErrorReducers', () => {
  test('It should reject if no transform set', () => {
    const err = new Error('Test')
    const accumulator = helpers.createAccumulator(
      {},
      {
        context: {
          error: helpers.createTransform([])
        }
      }
    )
    return ResolveEntity.resolveErrorReducers(
      err,
      accumulator,
      resolveTransform
    )
      .catch(reason => reason)
      .then(acc => {
        expect(acc).toBeInstanceOf(Error)
        expect(acc).toHaveProperty('message', 'Test')
      })
  })

  test('It should handle if transform set', () => {
    const err = new Error('Test')
    const accumulator = helpers.createAccumulator(
      {},
      {
        context: {
          error: helpers.createTransform((acc, next) => next(null, 'pass'))
        }
      }
    )
    return ResolveEntity.resolveErrorReducers(
      err,
      accumulator,
      resolveTransform
    ).then(acc => {
      expect(acc.value).toEqual('pass')
    })
  })
})

describe('ResolveEntity.createCurrentAccumulator', () => {
  let acc
  beforeAll(() => {
    const reducerEntity = ReducerEntity.create('hash:base')
    const accumulator = helpers.createAccumulator({
      foo: 'bar'
    })
    acc = ResolveEntity.createCurrentAccumulator(
      dataPoint,
      accumulator,
      reducerEntity
    )
  })
  test('It should set reducer property', () => {
    expect(acc).toHaveProperty('reducer.spec.id', 'hash:base')
  })
  test('It should context as the current Entity', () => {
    expect(acc).toHaveProperty('context.id', 'hash:base')
  })
  test('It should initialValue acc.value', () => {
    expect(acc).toHaveProperty('initialValue', {
      foo: 'bar'
    })
  })
  test('It should initialValue acc.value', () => {
    expect(acc).toHaveProperty('params', {
      base: true
    })
  })
})

describe('ResolveEntity.resolveMiddleware', () => {
  test('It should execute a middleware', () => {
    dataPoint.middleware.use('request:before', (acc, next) => {
      acc.value = 'bar'
      next(null)
    })

    const racc = helpers.createAccumulator('foo')
    return ResolveEntity.resolveMiddleware(
      dataPoint,
      'request:before',
      racc
    ).then(acc => {
      expect(acc.value).toEqual('bar')
    })
  })
})

describe('ResolveEntity.resolveEntity', () => {
  const defaultResolver = (acc, resolveTransform) => Promise.resolve(acc)

  const resolveEntity = (entityId, input, options, resolver) => {
    const racc = helpers.createAccumulator.call(null, input, options)
    const reducer = ReducerEntity.create(entityId)
    return ResolveEntity.resolveEntity(
      dataPoint,
      resolveTransform,
      racc,
      reducer,
      resolver || defaultResolver
    )
  }

  test('It should resolve entity', () => {
    return resolveEntity('hash:asIs', 'foo').then(acc => {
      expect(acc).toHaveProperty('value', 'foo')
    })
  })

  test('It should attach entityId to error', () => {
    const rejectResolver = () => Promise.reject(new Error('test'))
    return resolveEntity('hash:asIs', undefined, undefined, rejectResolver)
      .catch(error => error)
      .then(val => {
        expect(val).toHaveProperty('entityId', 'hash:asIs')
      })
  })

  test('It should log trace calls when set to true', () => {
    const consoleTime = console.time
    const consoleTimeEnd = console.timeEnd
    console.time = jest.fn()
    console.timeEnd = jest.fn()
    return resolveEntity('hash:asIs', 'foo', {
      trace: true
    }).then(acc => {
      expect(console.time).toBeCalled()
      expect(console.timeEnd).toBeCalled()
      expect(acc).toHaveProperty('value', 'foo')
      console.time = consoleTime
      console.timeEnd = consoleTimeEnd
    })
  })

  test('It should resolve through bypass', () => {
    dataPoint.middleware.use('hash:before', (acc, next) => {
      acc.resolve('bar')
      next(null)
    })
    return resolveEntity('hash:asIs', 'foo').then(acc => {
      expect(acc).toHaveProperty('value', 'bar')
    })
  })

  test('It should resolve through bypass', () => {
    dataPoint.middleware.use('hash:before', (acc, next) => {
      const err = new Error('test')
      throw err
    })
    return resolveEntity('hash:asIs', 'foo')
      .catch(err => err)
      .then(err => {
        expect(err).toHaveProperty('message', 'test')
      })
  })
})

describe('ResolveEntity.resolve', () => {
  const resolve = resolver => (entityId, input, options) => {
    const racc = helpers.createAccumulator.call(null, input, options)
    const reducer = ReducerEntity.create(entityId)
    return ResolveEntity.resolve(
      dataPoint,
      ReducerExpression.resolve,
      racc,
      reducer,
      resolver
    )
  }

  test('It should resolve as single entity', () => {
    const resolver = (acc, resolveTransform) => {
      const result = utils.set(acc, 'value', 'bar')
      return Promise.resolve(result)
    }
    return resolve(resolver)('hash:asIs', 'foo').then(acc => {
      expect(acc).toHaveProperty('value', 'bar')
    })
  })

  test('It should resolve as collection', () => {
    const resolver = (acc, resolveTransform) => {
      const result = utils.set(acc, 'value', 'bar')
      return Promise.resolve(result)
    }
    return resolve(resolver)('hash:asIs[]', ['foo']).then(acc => {
      expect(acc).toHaveProperty('value', ['bar'])
    })
  })
})
