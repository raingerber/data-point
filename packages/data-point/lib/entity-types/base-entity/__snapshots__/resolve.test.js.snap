// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolveEntity.resolveEntity inputType - throws error if inputType does not pass 1`] = `
[Error: Entity type check failed!
Expected type: number
Actual type: string
Input value: 'foo']
`;

exports[`ResolveEntity.resolveEntity outputType fails if error method catches typeCheck errors and returns bad value 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 2]
`;

exports[`ResolveEntity.resolveEntity outputType is bypassed if error throws error 1`] = `[Error: error from before method]`;

exports[`ResolveEntity.resolveEntity outputType throws error if after method returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if after middleware returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if before method returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if custom typeCheck fails 1`] = `[Error: custom type error]`;

exports[`ResolveEntity.resolveEntity outputType throws error if global after middleware returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if global before middleware returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if middleware before returns value that does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws error if value does not pass typeCheck 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;

exports[`ResolveEntity.resolveEntity outputType throws if error method does not return value with correct type 1`] = `
[Error: Entity type check failed!
Expected type: string
Actual type: number
Input value: 1]
`;
